source/opt/const_folding_rules.cpp

```diff
@@ -893,6 +897,124 @@ ConstantFoldingRule FoldFMix() {
+  if (const analysis::Integer* int_type = result_type->AsInteger()) {
+    if (int_type->width() == 32) {
+      if (int_type->IsSigned()) {
+        int32_t va = a->GetS32();
+        int32_t vb = b->GetS32();
+        return (va < vb ? a : b);
+      } else {
+        uint32_t va = a->GetU32();
+        uint32_t vb = b->GetU32();
+        return (va < vb ? a : b);
+      }
+    } else if (int_type->width() == 64) {
+      if (int_type->IsSigned()) {
+        int64_t va = a->GetS64();
+        int64_t vb = b->GetS64();
+        return (va < vb ? a : b);
+      } else {
+        uint64_t va = a->GetU64();
+        uint64_t vb = b->GetU64();
+        return (va < vb ? a : b);
+      }
+    }
+  } else if (const analysis::Float* float_type = result_type->AsFloat()) {
+    if (float_type->width() == 32) {
+      float va = a->GetFloat();
+      float vb = b->GetFloat();
+      return (va < vb ? a : b);
+    } else if (float_type->width() == 64) {
+      double va = a->GetDouble();
+      double vb = b->GetDouble();
+      return (va < vb ? a : b);
+    }
+  }
+  return nullptr;
+};
+
+BinaryScalarFoldingRule FoldMax =
+    [](const analysis::Type* result_type, const analysis::Constant* a,
+       const analysis::Constant* b,
+       analysis::ConstantManager*) -> const analysis::Constant* {
+  if (const analysis::Integer* int_type = result_type->AsInteger()) {
+    if (int_type->width() == 32) {
+      if (int_type->IsSigned()) {
+        int32_t va = a->GetS32();
+        int32_t vb = b->GetS32();
+        return (va > vb ? a : b);
+      } else {
+        uint32_t va = a->GetU32();
+        uint32_t vb = b->GetU32();
+        return (va > vb ? a : b);
+      }
+    } else if (int_type->width() == 64) {
+      if (int_type->IsSigned()) {
+        int64_t va = a->GetS64();
+        int64_t vb = b->GetS64();
+        return (va > vb ? a : b);
+      } else {
+        uint64_t va = a->GetU64();
+        uint64_t vb = b->GetU64();
+        return (va > vb ? a : b);
+      }
+    }
+  } else if (const analysis::Float* float_type = result_type->AsFloat()) {
+    if (float_type->width() == 32) {
+      float va = a->GetFloat();
+      float vb = b->GetFloat();
+      return (va > vb ? a : b);
+    } else if (float_type->width() == 64) {
+      double va = a->GetDouble();
+      double vb = b->GetDouble();
+      return (va > vb ? a : b);
+    }
+  }
+  return nullptr;
+};
+
+ConstantFoldingRule FoldClamp() {
+  return [](IRContext* context, Instruction* inst,
+            const std::vector<const analysis::Constant*>& constants)
+             -> const analysis::Constant* {
+    assert(inst->opcode() == SpvOpExtInst &&
+           "Expecting an extended instruction.");
+    assert(inst->GetSingleWordInOperand(0) ==
+               context->get_feature_mgr()->GetExtInstImportId_GLSLstd450() &&
+           "Expecting a GLSLstd450 extended instruction.");
+
+    // Make sure all Clamp operands are constants.
```

Comment:
This seems more restrictive than necessary.

```
clamp(1.0f, 0.0f, x) -> 1.0f
clamp(x, 2.0f, 1.0f) -> 1.0f
```

If this were a regular folding rule (not just a const rule) you could also fold:

```
clamp(1.0f, 2.0f, x) -> max(2.0f, x)
clamp(x, 1.0f, 2.0f) -> min(x, 1.0f)
```

Do you think it is worth handling these cases too?

Reply:
```
clamp(1.0f, 0.0f, x) -> 1.0f
```

This is not correct, This can be simplified to `min(1.0f, x)`, but no further.  However, I do not expect this to be very common.

```
clamp(x, 2.0f, 1.0f) -> 1.0f
```

This is undefined behaviour according to the instruction: "Result is undefined if minVal > maxVal." I could fold it, but it is not suppose to happen.

```
clamp(1.0f, 2.0f, x) -> max(2.0f, x)
```

We could add a folding rule for when the first two values are constants.  However the result is `min(2.0f, x)`.  Same as above.  I do not expect this to be common.

```
clamp(x, 1.0f, 2.0f) -> min(x, 1.0f)
```

This cannot be simplified.  If `x > 2`, then the result is `2`.  This will be the most common situation.

Reply:
I think I had my operands in the wrong order. Above I was using (incorrectly), min, val, max.

Reply:
I added folding rules for the first two.  We can leave the others for another PR if we see them showing up.